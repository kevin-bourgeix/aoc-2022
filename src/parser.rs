use std::fs;

use regex::Regex;

// Parses an aoc file into a vector of strings
// By defaults it splits on newlines, but it can be changed easily
pub fn parse_aoc_file(filename: &str, delimiter: Option<&str>) -> Vec<String> {
    let Ok(file) = fs::read_to_string(filename) else {
        panic!("Could not read file {filename}");
    };
    let split: Vec<&str> = file.split(delimiter.unwrap_or("\n")).collect();

    return split.iter().map(|s| s.to_string()).collect();
}

// Parses strings with the given parser function
pub fn parse_into_struct<T>(data: Vec<String>, parser: fn(String) -> T) -> Vec<T> {
    let mut result = Vec::new();
    for line in data {
        result.push(parser(line));
    }
    result
}

// Reads through a vector of lines while not matching a specific regex
pub fn parse_until_pattern(data: Vec<String>, pattern: &str) -> (Vec<String>, Vec<String>) {
    let Ok(re) = Regex::new(pattern) else {
        panic!("Could not parse pattern {pattern} at parse_until_pattern");
    };
    let mut index = 0;
    for line in &data {
        if re.is_match(line) {
            break;
        }
        index += 1;
    }
    (data[0..index].to_vec(), data[index..].to_vec())
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_snapshot;

    #[derive(Debug)]
    struct TestStruct {
        a: String,
        b: String,
    }

    #[test]
    fn test_parse_aoc_file_default() {
        let data = parse_aoc_file("./src/tests/numbers1.txt", None);
        insta::assert_debug_snapshot!(data);
    }

    #[test]
    fn test_parse_aoc_file_comma() {
        let data = parse_aoc_file("./src/tests/numbers2.txt", Some(","));
        insta::assert_debug_snapshot!(data);
    }

    #[test]
    fn test_parse_aoc_file_with_blank() {
        let data = parse_aoc_file("./src/tests/numbers3.txt", None);
        insta::assert_debug_snapshot!(data);
    }

    #[test]
    fn test_parse_into_struct() {
        let data = vec!["1,2".to_string(), "3,4".to_string()];
        let result = parse_into_struct(data, |s| {
            let split: Vec<&str> = s.split(',').collect();
            TestStruct {
                a: split[0].to_string(),
                b: split[1].to_string(),
            }
        });

        insta::assert_debug_snapshot!(result);
    }

    #[test]
    fn test_parse_into_simple() {
        let data = vec!["1".to_string(), "2".to_string()];
        let result = parse_into_struct(data, |s| s.parse::<i32>().unwrap());

        insta::assert_debug_snapshot!(result);
    }

    #[test]
    fn test_parse_until() {
        let data = vec!["aaaa", "aaaa", "aaaa", "bbbb", "bbbb", "cccc"]
            .iter()
            .map(|s| s.to_string())
            .collect();
        let (first, second) = parse_until_pattern(data, "bbbb");

        insta::assert_debug_snapshot!(vec![first, second]);
    }

    #[test]
    fn test_parse_empty() {
        let data = vec!["1", "", "2", "3"]
            .iter()
            .map(|s| s.to_string())
            .collect();
        let (first, second) = parse_until_pattern(data, r"^$");

        insta::assert_debug_snapshot!(vec![first, second]);
    }
}
