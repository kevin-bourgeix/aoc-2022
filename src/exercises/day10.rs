use core::panic;

use crate::parser::parse_aoc_file;

#[derive(Debug)]
enum Operation {
    Addx(i32),
    Noop,
}

fn parse_file(filename: &str) -> Vec<Operation> {
    let lines = parse_aoc_file(filename, None);
    lines
        .iter()
        .enumerate()
        .flat_map(|(i, l)| {
            let splits = l.split(' ').collect::<Vec<&str>>();
            let Some(op) = splits.first() else {
            panic!("Can't read operation {l} at line {i}");
        };
            let value = splits.get(1);
            if *op != "addx" && value.is_some() {
                panic!("Unknown operation {op} at line {i}");
            }

            match *op {
                "addx" => vec![
                    Operation::Noop,
                    Operation::Addx(value.unwrap().parse::<i32>().unwrap_or_else(|_| {
                        panic!(
                            "Can't parse value {} at line {}",
                            value.unwrap_or(&"NONE"),
                            i
                        )
                    })),
                ],
                "noop" => vec![Operation::Noop],
                _ => panic!("Unknown operation {op} at line {i}"),
            }
        })
        .collect()
}

fn run_operations(operations: Vec<Operation>, signal_start: u32, signal_step: u32) -> i32 {
    let mut iter = 0;
    let mut register: i32 = 1;

    operations.iter().enumerate().fold(0, |acc, (step, op)| {
        let mut current = acc;
        if (step + 1) as u32 == signal_start + iter * signal_step {
            current += register * (step + 1) as i32;
            iter += 1;
        }

        if let Operation::Addx(value) = op {
            register += value;
        }

        current
    })
}

// works only with parse_file_v2
fn run_operations_screen(
    operations: Vec<Operation>,
    signal_size: i32,
    screen_height: u32,
) -> Vec<Vec<bool>> {
    let mut register: i32 = 1;
    let mut screen: Vec<Vec<bool>> = vec![];
    let mut ops = operations.iter().cycle();

    for _ in 0..screen_height {
        let line: Vec<bool> = (0..signal_size)
            .map(|i| {
                let mut value = false;

                // safe unwrap as ops is a cycle
                let op = ops.next().unwrap();
                if i == register - 1 || i == register || i == register + 1 {
                    value = true;
                }
                if let Operation::Addx(v) = op {
                    register += v;
                }
                value
            })
            .collect();
        screen.push(line);
    }

    screen
}

pub fn day_10_1(filename: &str) -> i32 {
    let operations = parse_file(filename);
    run_operations(operations, 20, 40)
}

pub fn day_10_2(filename: &str) -> Vec<Vec<bool>> {
    let operations = parse_file(filename);
    run_operations_screen(operations, 40, 6)
}

#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;

    use super::*;

    #[test]
    fn test_exec_op() {
        let operations = parse_file("src/files/day10_1.test");
        let total = run_operations(operations, 20, 40);
        assert_eq!(total, 13140);
    }

    #[test]
    fn test_exec_op_screen() {
        let operations = parse_file("src/files/day10_1.test");
        let screen = run_operations_screen(operations, 40, 6);
        assert_debug_snapshot!(screen);
    }
}
