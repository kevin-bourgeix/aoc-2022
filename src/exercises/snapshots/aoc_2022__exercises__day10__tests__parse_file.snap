---
source: src/exercises/day10.rs
assertion_line: 49
expression: operations
---
[
    Addx(
        15,
    ),
    Addx(
        -11,
    ),
    Addx(
        6,
    ),
    Addx(
        -3,
    ),
    Addx(
        5,
    ),
    Addx(
        -1,
    ),
    Addx(
        -8,
    ),
    Addx(
        13,
    ),
    Addx(
        4,
    ),
    Noop,
    Addx(
        -1,
    ),
    Addx(
        5,
    ),
    Addx(
        -1,
    ),
    Addx(
        5,
    ),
    Addx(
        -1,
    ),
    Addx(
        5,
    ),
    Addx(
        -1,
    ),
    Addx(
        5,
    ),
    Addx(
        -1,
    ),
    Addx(
        -35,
    ),
    Addx(
        1,
    ),
    Addx(
        24,
    ),
    Addx(
        -19,
    ),
    Addx(
        1,
    ),
    Addx(
        16,
    ),
    Addx(
        -11,
    ),
    Noop,
    Noop,
    Addx(
        21,
    ),
    Addx(
        -15,
    ),
    Noop,
    Noop,
    Addx(
        -3,
    ),
    Addx(
        9,
    ),
    Addx(
        1,
    ),
    Addx(
        -3,
    ),
    Addx(
        8,
    ),
    Addx(
        1,
    ),
    Addx(
        5,
    ),
    Noop,
    Noop,
    Noop,
    Noop,
    Noop,
    Addx(
        -36,
    ),
    Noop,
    Addx(
        1,
    ),
    Addx(
        7,
    ),
    Noop,
    Noop,
    Noop,
    Addx(
        2,
    ),
    Addx(
        6,
    ),
    Noop,
    Noop,
    Noop,
    Noop,
    Noop,
    Addx(
        1,
    ),
    Noop,
    Noop,
    Addx(
        7,
    ),
    Addx(
        1,
    ),
    Noop,
    Addx(
        -13,
    ),
    Addx(
        13,
    ),
    Addx(
        7,
    ),
    Noop,
    Addx(
        1,
    ),
    Addx(
        -33,
    ),
    Noop,
    Noop,
    Noop,
    Addx(
        2,
    ),
    Noop,
    Noop,
    Noop,
    Addx(
        8,
    ),
    Noop,
    Addx(
        -1,
    ),
    Addx(
        2,
    ),
    Addx(
        1,
    ),
    Noop,
    Addx(
        17,
    ),
    Addx(
        -9,
    ),
    Addx(
        1,
    ),
    Addx(
        1,
    ),
    Addx(
        -3,
    ),
    Addx(
        11,
    ),
    Noop,
    Noop,
    Addx(
        1,
    ),
    Noop,
    Addx(
        1,
    ),
    Noop,
    Noop,
    Addx(
        -13,
    ),
    Addx(
        -19,
    ),
    Addx(
        1,
    ),
    Addx(
        3,
    ),
    Addx(
        26,
    ),
    Addx(
        -30,
    ),
    Addx(
        12,
    ),
    Addx(
        -1,
    ),
    Addx(
        3,
    ),
    Addx(
        1,
    ),
    Noop,
    Noop,
    Noop,
    Addx(
        -9,
    ),
    Addx(
        18,
    ),
    Addx(
        1,
    ),
    Addx(
        2,
    ),
    Noop,
    Noop,
    Addx(
        9,
    ),
    Noop,
    Noop,
    Noop,
    Addx(
        -1,
    ),
    Addx(
        2,
    ),
    Addx(
        -37,
    ),
    Addx(
        1,
    ),
    Addx(
        3,
    ),
    Noop,
    Addx(
        15,
    ),
    Addx(
        -21,
    ),
    Addx(
        22,
    ),
    Addx(
        -6,
    ),
    Addx(
        1,
    ),
    Noop,
    Addx(
        2,
    ),
    Addx(
        1,
    ),
    Noop,
    Addx(
        -10,
    ),
    Noop,
    Noop,
    Addx(
        20,
    ),
    Addx(
        1,
    ),
    Addx(
        2,
    ),
    Addx(
        2,
    ),
    Addx(
        -6,
    ),
    Addx(
        -11,
    ),
    Noop,
    Noop,
    Noop,
]
